### PR TITLE
1.65.0 release notes

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
r? @cuviper (since you're writing the blog)

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
cc @rust-lang/release

File: RELEASES.md [hunk 1]
```diff
@@ -1,3 +1,91 @@
...
+- [Uninitialized integers, floats, and raw pointers are now considered immediate UB](https://github.com/rust-lang/rust/pull/98919/)
+  Usage of `MaybeUninit` is the correct way to work with uninitialized memory.
+- [Stabilize raw-dylib for Windows x86_64, aarch64, and thumb7a](https://github.com/rust-lang/rust/pull/99916/)
+- [Do not allow `Drop` impl on foreign ADTs](https://github.com/rust-lang/rust/pull/99576/)
+
+Compiler
+--------
+- [Stabilize -Csplit-debuginfo on Linux](https://github.com/rust-lang/rust/pull/98051/)
+- [Use niche-filling optimization even when multiple variants have data](https://github.com/rust-lang/rust/pull/94075/)
+- [Associated type projects are now verified to be well-formed prior to resolving the underlying type](https://github.com/rust-lang/rust/pull/99217/#issuecomment-1209365630)
```
Comment (memoryruins) [inline RELEASES.md:20]:
```suggestion
- [Associated type projections are now verified to be well-formed prior to resolving the underlying type](https://github.com/rust-lang/rust/pull/99217/#issuecomment-1209365630)
```

File: RELEASES.md [hunk 2]
```diff
@@ -1,3 +1,91 @@
...
+----------------
+
+These changes do not affect any public interfaces of Rust, but they represent
+significant improvements to the performance or internals of rustc and related
+tools.
+
+- [Add `x.sh` and `x.ps1` shell scripts](https://github.com/rust-lang/rust/pull/99992/)
+- [compiletest: use target cfg instead of hard-coded tables](https://github.com/rust-lang/rust/pull/100260/)
+- [Use object instead of LLVM for reading bitcode from rlibs](https://github.com/rust-lang/rust/pull/98100/)
+
```
Comment (wesleywiser) [inline RELEASES.md:88]: It is probably worth mentioning that this is the first release with MIR inlining enabled for compiles with optimization enabled leading to [3-10% throughput improvements](https://perf.rust-lang.org/compare.html?start=aedf78e56b2279cc869962feac5153b6ba7001ed&end=0075bb4fad68e64b6d1be06bf2db366c30bc75e1&stat=instructions:u) on real-world crates (#91743). 

Note, that PR is milestoned at 1.64 but we disabled the optimization during beta to allow for more testing. Thus 1.65 is the first release with the changes enabled.

File: RELEASES.md [hunk 3]
```diff
@@ -1,3 +1,91 @@
...
+Cargo
+-----
+
+- [Apply GitHub fast path even for partial hashes](https://github.com/rust-lang/cargo/pull/10807/)
+- [Do not add home bin path to PATH if it's already there](https://github.com/rust-lang/cargo/pull/11023/)
+- [Take priority into account within the pending queue](https://github.com/rust-lang/cargo/pull/11032/)
+  This slightly optimizes job scheduling by Cargo, with typically small improvements on larger crate graph builds.
+
+Compatibility Notes
```
Comment (cuviper) [inline RELEASES.md:67]: We should add a note for #99880, which is being backported.

File: RELEASES.md [hunk 4]
```diff
@@ -1,3 +1,91 @@
...
+- [Document no support for writing to non-blocking stdio/stderr](https://github.com/rust-lang/rust/pull/101416/)
+
+Stabilized APIs
+---------------
+
+- [`std::backtrace::Backtrace`](https://doc.rust-lang.org/stable/std/backtrace/struct.Backtrace.html)
+- [`Bound::as_ref`](https://doc.rust-lang.org/stable/std/ops/enum.Bound.html#method.as_ref)
+- [`std::io::read_to_string`](https://doc.rust-lang.org/stable/std/io/fn.read_to_string.html)
+- [`<*const T>::cast_mut`](https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.cast_mut)
+- [`<*mut T>::cast_const`](https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.cast_mut)
```
Comment (tanriol) [inline RELEASES.md:52]: Still links to `cast_mut`, not `cast_const`

Comment (pthariensflame) [reply RELEASES.md:52]:
```suggestion
- [`<*mut T>::cast_const`](https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.cast_const)
```

**Review by ogoffart — APPROVED** [review]

File: RELEASES.md [hunk 5]
```diff
@@ -1,3 +1,91 @@
+Version 1.65.0 (2022-11-03)
+==========================
+
+Language
+--------
+- [Error on `as` casts of enums with `#[non_exhaustive]`](https://github.com/rust-lang/rust/pull/92744/)
+- [Stabilize `let else`](https://github.com/rust-lang/rust/pull/93628/)
+- [Stabilize generic associated types (GATs)](https://github.com/rust-lang/rust/pull/96709/)
+- [Add lints `let_underscore_drop`, `let_underscore_lock`, and `let_underscore_must_use` from Clippy](https://github.com/rust-lang/rust/pull/97739/)
+- [Stabilize `break`ing from arbitrary labeled blocks](https://github.com/rust-lang/rust/pull/99332/)
```
Comment (joshtriplett) [inline RELEASES.md:10]:
```suggestion
- [Stabilize `break`ing from arbitrary labeled blocks ("label-break-value")](https://github.com/rust-lang/rust/pull/99332/)
```

File: RELEASES.md [hunk 6]
```diff
@@ -1,3 +1,91 @@
...
+- [Add powerpc64-unknown-openbsd and riscv64-unknown-openbsd as tier 3 targets](https://github.com/rust-lang/rust/pull/101025/)
+  - Refer to Rust's [platform support page][platform-support-doc] for more
+    information on Rust's tiered platform support.
+
+Libraries
+---------
+
+- [Don't generate `PartialEq::ne` in derive(PartialEq)](https://github.com/rust-lang/rust/pull/98655/)
+- [Windows RNG: Use `BCRYPT_RNG_ALG_HANDLE` by default](https://github.com/rust-lang/rust/pull/101325/)
+- [Forbid mixing `System` with direct sytem allocator calls](https://github.com/rust-lang/rust/pull/101394/)
```
Comment (joshtriplett) [inline RELEASES.md:42]:
```suggestion
- [Forbid mixing `System` with direct system allocator calls](https://github.com/rust-lang/rust/pull/101394/)
```

File: RELEASES.md [hunk 7]
```diff
@@ -1,3 +1,95 @@
...
+
+Language
+--------
+- [Error on `as` casts of enums with `#[non_exhaustive]`](https://github.com/rust-lang/rust/pull/92744/)
+- [Stabilize `let else`](https://github.com/rust-lang/rust/pull/93628/)
+- [Stabilize generic associated types (GATs)](https://github.com/rust-lang/rust/pull/96709/)
+- [Add lints `let_underscore_drop`, `let_underscore_lock`, and `let_underscore_must_use` from Clippy](https://github.com/rust-lang/rust/pull/97739/)
+- [Stabilize `break`ing from arbitrary labeled blocks ("label-break-value")](https://github.com/rust-lang/rust/pull/99332/)
+- [Uninitialized integers, floats, and raw pointers are now considered immediate UB](https://github.com/rust-lang/rust/pull/98919/)
+  Usage of `MaybeUninit` is the correct way to work with uninitialized memory.
```
Comment (wesleywiser) [inline RELEASES.md:12]: I feel like this should link to something. `MaybeUninit` is already stable so it's a bit unclear what this means.

Comment (wesleywiser) [reply RELEASES.md:12]: Oh, sorry! I read this as a separate bullet point from the previous line and not as the second sentence in the same point.

File: RELEASES.md [hunk 8]
```diff
@@ -1,3 +1,95 @@
...
+- [Stabilize raw-dylib for Windows x86_64, aarch64, and thumbv7a](https://github.com/rust-lang/rust/pull/99916/)
+- [Do not allow `Drop` impl on foreign ADTs](https://github.com/rust-lang/rust/pull/99576/)
+
+Compiler
+--------
+- [Stabilize -Csplit-debuginfo on Linux](https://github.com/rust-lang/rust/pull/98051/)
+- [Use niche-filling optimization even when multiple variants have data](https://github.com/rust-lang/rust/pull/94075/)
```
Comment (samanpa) [inline RELEASES.md:19]: Should this be considered a language change?

Comment (samanpa) [reply RELEASES.md:19]: Thanks that makes sense. This change is probably worth highlighting the release announcement.

**Comment by inquisitivecrystal** [comment]
Just to make sure this is on all the right people's minds, #103450 is going to need a compatibility note. I don't know what the cutoff for putting something in the blog post is, but it may be worth considering it, especially given that this is being merged so close to release.

**Comment by Mark-Simulacrum** [comment]
https://github.com/rust-lang/rust/pull/102737 will need a release notes item for the intentional breakage.

**Comment by Mark-Simulacrum** [comment]
Squashed commits again; @bors r+ p=1

File: RELEASES.md [hunk 9]
```diff
@@ -1,3 +1,106 @@
+Version 1.65.0 (2022-11-03)
+==========================
+
+Language
+--------
+- [Error on `as` casts of enums with `#[non_exhaustive]`](https://github.com/rust-lang/rust/pull/92744/)
```
Comment (joshtriplett) [inline RELEASES.md:6]:
```suggestion
- [Error on `as` casts of enums with `#[non_exhaustive]` variants](https://github.com/rust-lang/rust/pull/92744/)
```

AFAICT the check is for enum variants that are non_exhaustive, not if the whole enum is non_exhaustive.

File: RELEASES.md [hunk 10]
```diff
@@ -1,3 +1,106 @@
...
+- [Stabilize `let else`](https://github.com/rust-lang/rust/pull/93628/)
+- [Stabilize generic associated types (GATs)](https://github.com/rust-lang/rust/pull/96709/)
+- [Add lints `let_underscore_drop`, `let_underscore_lock`, and `let_underscore_must_use` from Clippy](https://github.com/rust-lang/rust/pull/97739/)
+- [Stabilize `break`ing from arbitrary labeled blocks ("label-break-value")](https://github.com/rust-lang/rust/pull/99332/)
+- [Uninitialized integers, floats, and raw pointers are now considered immediate UB](https://github.com/rust-lang/rust/pull/98919/)
```
Comment (joshtriplett) [inline RELEASES.md:11]:
```suggestion
- [Uninitialized integers, floats, and raw pointers are now considered immediate UB](https://github.com/rust-lang/rust/pull/98919/).
```

File: RELEASES.md [hunk 11]
```diff
@@ -1,3 +1,106 @@
...
+- [`<*const T>::offset_from`](https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.offset_from)
+- [`<*mut T>::offset_from`](https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.offset_from)
+
+Cargo
+-----
+
+- [Apply GitHub fast path even for partial hashes](https://github.com/rust-lang/cargo/pull/10807/)
+- [Do not add home bin path to PATH if it's already there](https://github.com/rust-lang/cargo/pull/11023/)
+- [Take priority into account within the pending queue](https://github.com/rust-lang/cargo/pull/11032/)
+  This slightly optimizes job scheduling by Cargo, with typically small improvements on larger crate graph builds.
```
Comment (joshtriplett) [inline RELEASES.md:67]:
```suggestion
- [Take priority into account within the pending queue](https://github.com/rust-lang/cargo/pull/11032/).
  This slightly optimizes job scheduling by Cargo, with typically small improvements on larger crate graph builds.
```

File: RELEASES.md [hunk 12]
```diff
@@ -1,3 +1,106 @@
...
+
+Compatibility Notes
+-------------------
+
+- [`std::layout::Layout` size must not overflow `isize::MAX` when rounded up to `align`](https://github.com/rust-lang/rust/pull/95295)
+  This also changes the safety conditions on `Layout::from_size_align_unchecked`.
+- [`PollFn` now only implements `Unpin` if the closure is `Unpin`](https://github.com/rust-lang/rust/pull/102737)
```
Comment (joshtriplett) [inline RELEASES.md:74]:
```suggestion
- [`PollFn` now only implements `Unpin` if the closure is `Unpin`](https://github.com/rust-lang/rust/pull/102737).
```

File: RELEASES.md [hunk 13]
```diff
@@ -1,3 +1,106 @@
...
+
+- [Apply GitHub fast path even for partial hashes](https://github.com/rust-lang/cargo/pull/10807/)
+- [Do not add home bin path to PATH if it's already there](https://github.com/rust-lang/cargo/pull/11023/)
+- [Take priority into account within the pending queue](https://github.com/rust-lang/cargo/pull/11032/)
+  This slightly optimizes job scheduling by Cargo, with typically small improvements on larger crate graph builds.
+
+Compatibility Notes
+-------------------
+
+- [`std::layout::Layout` size must not overflow `isize::MAX` when rounded up to `align`](https://github.com/rust-lang/rust/pull/95295)
```
Comment (joshtriplett) [inline RELEASES.md:72]:
```suggestion
- [`std::layout::Layout` size must not overflow `isize::MAX` when rounded up to `align`](https://github.com/rust-lang/rust/pull/95295).
```

**Comment by Mark-Simulacrum** [comment]
Squashed commits again -- thanks @joshtriplett.

@bors r+ p=1

**Pull request merged**